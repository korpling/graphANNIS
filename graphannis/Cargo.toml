[package]
authors = ["Thomas Krause <thomaskrause@posteo.de>"]
build = "build.rs"
description = "This is a new backend implementation of the ANNIS linguistic search and visualization system."
edition = "2021"
license = "Apache-2.0"
name = "graphannis"
readme = "../README.md"
repository = "https://github.com/korpling/graphANNIS"
version = "3.6.0"

[lib]
crate-type = ["lib"]

[build-dependencies]
csv = "1.1"
file_diff = "1"
lalrpop = { version = "0.20", default-features = false, features = [
    "lexer",
    "unicode",
] }
regex = "1"

[dependencies]
boolean_expression = "0.4"
csv = "1"
fs2 = "0.4"
graphannis-core = { path = "../core/", version = "^3" }
itertools = "0.10"
lalrpop-util = { version = "0.20", features = ["lexer"] }
lazy_static = "1.4"
libc = "0.2"
linked-hash-map = "0.5"
log = "0.4"
lru = "0.7"
memory-stats = "1.1.0"
page_size = "0.4"
percent-encoding = "2.1"
rand = { version = "0.8", features = ["small_rng"] }
rayon = { version = "1.3", default-features = false }
regex = "1"
regex-syntax = "0.8"
rustc-hash = "1.0"
serde = { version = "1.0", features = ["rc"] }
serde_derive = "1.0"
smallvec = "1.6"
smartstring = { version = "1", features = ["serde"] }
strum = "0.21"
strum_macros = "0.21"
sys-info = "0.9"
tempfile = "3"
thiserror = "1"
toml = "0.8"
transient-btree-index = "0.5"
zip = "0.6.4"

[dev-dependencies]
assert_matches = "1.5.0"
criterion = "0.5"
fake = "2.2"
insta = "1.34.0"
pretty_assertions = "1.3"
same-file = "1.0.6"
serial_test = "2"

[[bench]]
harness = false
name = "graphannis"
