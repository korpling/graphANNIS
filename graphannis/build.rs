use file_diff::diff;

use regex::Regex;
use std::path::PathBuf;

/// Take the CSV file with the queries and add a test case for each query whose
/// corpora exist
fn create_search_tests() -> Option<()> {
    let out_dir = std::env::var("OUT_DIR").unwrap();
    let destination = std::path::Path::new(&out_dir).join("searchtest.rs");
    let destintation_tmp = std::path::Path::new(&out_dir).join("searchtest.rs.tmp");
    let mut f = std::fs::File::create(&destintation_tmp).unwrap();

    let query_file = PathBuf::from(if let Ok(path) = std::env::var("ANNIS4_TEST_QUERIES") {
        path
    } else {
        String::from("tests/searchtest_queries.csv")
    });
    println!("cargo:rerun-if-changed={}", query_file.to_string_lossy());
    let invalid_chars = Regex::new(r"[^A-Za-z0-9_]").unwrap();

    if query_file.is_file() {
        let mut reader = csv::Reader::from_path(query_file).ok()?;
        for row in reader.records() {
            use std::io::Write;
            let row = row.ok()?;
            let name = row.get(0)?;
            let name_escaped = invalid_chars.replace_all(name, "_");
            let aql = row.get(1)?;
            let corpus = row.get(2)?;
            let corpus_escaped = invalid_chars.replace_all(corpus, "_");
            let count = row.get(3)?.trim().parse::<i64>().ok()?;
            // output test case
            write!(
                f,
                "
#[ignore]
#[test]
#[allow(non_snake_case)]
fn search_{corpus_escaped}_{name_escaped}() {{
let aql = r###\"{aql}\"###;
if let Some(cs) = CORPUS_STORAGE.as_ref() {{
    let search_count = {{ 
        let search_query = graphannis::corpusstorage::SearchQuery {{
            query: aql,
            corpus_names: &[\"{corpus}\"],
            query_language: QueryLanguage::AQL,
            timeout: None,
        }};
        cs.count(search_query).unwrap_or(0)
    }};
    assert_eq!(
        {count}, search_count,
        \"Query '{{}}' ({name}) on corpus {corpus} should have had count {count} but was {{}}.\", 
        aql, search_count
    );
}}
}}

",
                name = name,
                name_escaped = name_escaped,
                corpus = corpus,
                corpus_escaped = corpus_escaped,
                aql = aql,
                count = count
            )
            .ok()?;
        }
    }

    // if the new temporary file is different to the existing one, replace them
    if !diff(
        &destination.to_string_lossy(),
        &destintation_tmp.to_string_lossy(),
    ) {
        std::fs::copy(destintation_tmp, destination).ok()?;
    }

    Some(())
}

fn create_parser() {
    println!("cargo:rerun-if-changed=src/annis/db/aql/parser.lalrpop");
    lalrpop::process_root().unwrap();
}

fn main() {
    create_parser();
    create_search_tests().unwrap();
}
