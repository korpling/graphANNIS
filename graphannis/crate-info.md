graphANNIS
==========

This is a new backend implementation of the ANNIS linguistic search and visualization system (http://corpus-tools.org/annis/).

**Only a sub-set of the ANNIS Query Langugage (AQL) from ANNIS version 3 (based on PostgreSQL) is supported yet.**
More operators can be added in the future, but the ones missing are the ones which have been used less frequent.
There is a tutorial in the Developer Guide on how to embedd graphANNIS in your own application.

The basic design ideas and data models are described in detail in the PhD-thesis ["ANNIS: A graph-based query system for deeply annotated text corpora"](https://doi.org/10.18452/19659). The thesis describes a prototype implementation in C++ and not Rust, but the design ideas are the same.
Noteable differences/enhancements compared to the thesis are:

- Graph storages implement querying inverse edges and finding reachable nodes based on them: this allows to implement inverse operators (e.g. for precedence) and switching operands in situations where it was not possible before.
- The data model has been simplified: the inverse coverage component and inverse edges in the left-/right-most token component have been removed.
- Additional query language features are now supported.