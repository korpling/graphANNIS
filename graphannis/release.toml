# Replace file contents only once
pre-release-replacements = [
  {file = "../CHANGELOG.md", search = "## \\[Unreleased\\]", replace = "## [Unreleased]\n\n## [{{version}}] - {{date}}"},
  {file = "../CITATION.cff", search = "^version: [0-9.]+", replace = "version: {{version}}"},
  {file = "../CITATION.cff", search = "date-released: [0-9\\-]+", replace = "date-released: {{date}}"},
  {file = "../docs/src/embed-rust.md", search = "graphannis = \"[0-9.]+\"", replace = "graphannis = \"{{version}}\""},
]
tag-name = "v{{version}}"
