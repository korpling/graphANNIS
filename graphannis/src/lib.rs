//! This is a graph-based linguistic corpus query system which implements the ANNIS Query Language (AQL).
//! The main entry point to the API is the [CorpusStorage](struct.CorpusStorage.html) struct which allows to manage and query a database of corpora.

#![deny(
    clippy::panic,
    clippy::panic_in_result_fn,
    clippy::expect_used,
    clippy::exit,
    clippy::todo,
    clippy::unwrap_in_result
)]

#[macro_use]
extern crate log;

#[macro_use]
extern crate serde_derive;

#[macro_use]
extern crate lazy_static;

#[macro_use]
extern crate lalrpop_util;

mod annis;

pub use crate::annis::db::corpusstorage::CorpusStorage;

/// Types that are used by the `CorpusStorage` API.
pub mod corpusstorage {
    pub use crate::annis::db::corpusstorage::SearchQuery;
    pub use crate::annis::db::corpusstorage::{
        CacheStrategy, CorpusInfo, ExportFormat, FrequencyDefEntry, GraphStorageInfo, ImportFormat,
        LoadStatus, QueryLanguage, ResultOrder,
    };
    pub use crate::annis::types::{
        CountExtra, FrequencyTable, FrequencyTableRow, QueryAttributeDescription,
    };
}

pub use graphannis_core::graph::update;

pub use graphannis_core::graph::Graph;

/// A specialization of the [`Graph`], using components needed to represent and query corpus annotation graphs.
pub type AnnotationGraph =
    graphannis_core::graph::Graph<annis::db::aql::model::AnnotationComponentType>;

/// Types that are used by the `Graph` API.
pub mod graph {
    pub use graphannis_core::annostorage::AnnotationStorage;
    pub use graphannis_core::annostorage::Match;
    pub use graphannis_core::annostorage::MatchGroup;
    pub use graphannis_core::graph::storage::GraphStatistic;
    pub use graphannis_core::graph::storage::{EdgeContainer, GraphStorage, WriteableGraphStorage};
    pub use graphannis_core::types::{AnnoKey, Annotation, Component, Edge, NodeID};
}

/// Types that define the annotation graph model.
pub mod model {
    pub use crate::annis::db::aql::model::AnnotationComponentType;
    pub type AnnotationComponent =
        graphannis_core::types::Component<crate::model::AnnotationComponentType>;
}

/// Helper functions to execute AQL queries on an [`AnnotationGraph`].
pub mod aql {
    pub use crate::annis::db::aql::disjunction::Disjunction;
    pub use crate::annis::db::aql::execute_query_on_graph;
    pub use crate::annis::db::aql::parse;
}

/// Contains the graphANNIS-specific error types.
pub mod errors {
    pub use crate::annis::errors::*;
}

/// Utility functions.
pub mod util {
    pub use crate::annis::util::SearchDef;
    pub use crate::annis::util::get_queries_from_csv;
    pub use crate::annis::util::node_names_from_match;
}
