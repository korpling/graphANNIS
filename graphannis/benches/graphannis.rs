#[macro_use]
extern crate criterion;
#[macro_use]
extern crate lazy_static;
extern crate graphannis;
extern crate rand;
extern crate rustc_hash;

use criterion::Criterion;
use fake::Fake;
use fake::faker::name::raw::*;
use fake::locales::*;
use graphannis::CorpusStorage;
use graphannis::corpusstorage::ResultOrder;
use graphannis::corpusstorage::{QueryLanguage, SearchQuery};
use graphannis::update::{GraphUpdate, UpdateEvent};
use std::collections::HashSet;
use std::path::Path;
use std::path::PathBuf;

lazy_static! {

static ref CORPUS_STORAGE : Option<CorpusStorage> = {
        let db_dir = PathBuf::from(if let Ok(path) = std::env::var("ANNIS4_TEST_DATA") {
            path
        }
        else if Path::new("data").is_dir() {
            String::from("data")
        } else {
            String::from("../data")
        });

        // only execute the test if the directory exists
        if db_dir.exists() && db_dir.is_dir() {
            CorpusStorage::with_auto_cache_size(&db_dir, false).ok()
        } else {
            None
        }
    };
}

const TOK_COUNT: usize = 100_000;

fn find_all_nouns_gum(bench: &mut Criterion) {
    if CORPUS_STORAGE.is_none() {
        return;
    }

    let cs = CORPUS_STORAGE.as_ref().unwrap();

    let corpora = cs.list();
    if let Ok(corpora) = corpora {
        let corpora: HashSet<String> = corpora.into_iter().map(|c| c.name).collect();
        // ignore if corpus does not exist
        if corpora.contains("GUM") {
            cs.preload("GUM").unwrap();
        } else {
            return;
        }
    }

    bench.bench_function("find_all_nouns_gum", move |b| {
        cs.preload("GUM").unwrap();
        b.iter(|| {
            let query = SearchQuery {
                corpus_names: &["GUM"],
                query: "pos=\"NN\"",
                query_language: QueryLanguage::AQL,
                timeout: None,
            };
            let f = cs.find(query, 0, None, ResultOrder::Normal);
            assert!(f.is_ok());
        })
    });
}

fn find_first_ten_nouns_gum(bench: &mut Criterion) {
    if CORPUS_STORAGE.is_none() {
        return;
    }

    let cs = CORPUS_STORAGE.as_ref().unwrap();

    let corpora = cs.list();
    if let Ok(corpora) = corpora {
        let corpora: HashSet<String> = corpora.into_iter().map(|c| c.name).collect();
        // ignore if corpus does not exist
        if corpora.contains("GUM") {
            cs.preload("GUM").unwrap();
        } else {
            return;
        }
    }

    bench.bench_function("find_first_ten_nouns_gum", move |b| {
        cs.preload("GUM").unwrap();
        b.iter(|| {
            let query = SearchQuery {
                corpus_names: &["GUM"],
                query: "pos=\"NN\"",
                query_language: QueryLanguage::AQL,
                timeout: None,
            };
            let f = cs.find(query, 0, Some(10), ResultOrder::Normal);
            assert!(f.is_ok());
        })
    });
}

fn count_extra_coref_anno(bench: &mut Criterion) {
    if CORPUS_STORAGE.is_none() {
        return;
    }

    let cs = CORPUS_STORAGE.as_ref().unwrap();

    let corpora = cs.list();
    if let Ok(corpora) = corpora {
        let corpora: HashSet<String> = corpora.into_iter().map(|c| c.name).collect();
        // ignore if corpus does not exist
        if !corpora.contains("GUM") {
            return;
        }
    }

    bench.bench_function("count_extra_coref_anno", move |b| {
        cs.preload("GUM").unwrap();
        let query = SearchQuery {
            corpus_names: &["GUM"],
            query: "infstat=\"giv\" ->coref[type=\"coref\"] entity=\"person\"",
            query_language: QueryLanguage::AQL,
            timeout: None,
        };
        b.iter(move || {
            let r = cs.count_extra(query.clone());
            assert!(r.is_ok());
        })
    });
}

fn count_extra_node(bench: &mut Criterion) {
    if CORPUS_STORAGE.is_none() {
        return;
    }

    let cs = CORPUS_STORAGE.as_ref().unwrap();

    let corpora = cs.list();
    if let Ok(corpora) = corpora {
        let corpora: HashSet<String> = corpora.into_iter().map(|c| c.name).collect();
        // ignore if corpus does not exist
        if !corpora.contains("GUM") {
            return;
        }
    }

    bench.bench_function("count_extra_node", move |b| {
        cs.preload("GUM").unwrap();
        let query = SearchQuery {
            corpus_names: &["GUM"],
            query: "node",
            query_language: QueryLanguage::AQL,
            timeout: None,
        };

        b.iter(move || {
            let r = cs.count_extra(query.clone());
            assert!(r.is_ok());
        })
    });
}

fn find_first_ten_token_gum(bench: &mut Criterion) {
    if CORPUS_STORAGE.is_none() {
        return;
    }

    let cs = CORPUS_STORAGE.as_ref().unwrap();

    let corpora = cs.list();
    if let Ok(corpora) = corpora {
        let corpora: HashSet<String> = corpora.into_iter().map(|c| c.name).collect();
        // ignore if corpus does not exist
        if corpora.contains("GUM") {
            cs.preload("GUM").unwrap();
        } else {
            return;
        }
    }

    bench.bench_function("find_first_ten_token_gum", move |b| {
        cs.preload("GUM").unwrap();
        b.iter(|| {
            let query = SearchQuery {
                corpus_names: &["GUM"],
                query: "tok",
                query_language: QueryLanguage::AQL,
                timeout: None,
            };
            let f = cs.find(query, 0, Some(10), ResultOrder::Normal);
            assert!(f.is_ok());
        })
    });
}

fn deserialize_gum(bench: &mut Criterion) {
    if CORPUS_STORAGE.is_none() {
        return;
    }

    let cs = CORPUS_STORAGE.as_ref().unwrap();
    let corpora = cs.list();
    if let Ok(corpora) = corpora {
        let corpora: HashSet<String> = corpora.into_iter().map(|c| c.name).collect();
        // ignore if corpus does not exist
        if corpora.contains("GUM") {
            cs.preload("GUM").unwrap();
        } else {
            return;
        }
    }

    bench.bench_function("deserialize_gum", move |b| {
        b.iter(|| {
            cs.unload("GUM").unwrap();
            cs.preload("GUM").unwrap();
        });
    });
}

/// Creates graph updates for adding some token (first result) and deleting them
/// again (second result item).
fn create_example_update() -> (GraphUpdate, GraphUpdate) {
    // Create a set of graph updates to apply
    let mut updates_add = GraphUpdate::default();

    // Generate a lot of tokens made of fake strings (using names)
    let mut token_names: Vec<String> = Vec::with_capacity(TOK_COUNT);
    let mut previous_token_name = None;
    for i in 0..TOK_COUNT {
        let node_name = format!("n{}", i);

        let t: &str = LastName(EN).fake();

        token_names.push(node_name.clone());

        // Create token node
        updates_add
            .add_event(UpdateEvent::AddNode {
                node_name: node_name.clone(),
                node_type: "node".to_string(),
            })
            .unwrap();
        updates_add
            .add_event(UpdateEvent::AddNodeLabel {
                node_name: node_name.clone(),
                anno_ns: "annis".to_string(),
                anno_name: "tok".to_string(),
                anno_value: t.to_string(),
            })
            .unwrap();

        // add the order relation
        if let Some(previous_token_name) = previous_token_name {
            updates_add
                .add_event(UpdateEvent::AddEdge {
                    source_node: previous_token_name,
                    target_node: node_name.clone(),
                    layer: "annis".to_string(),
                    component_type: "Ordering".to_string(),
                    component_name: "".to_string(),
                })
                .unwrap();
        }

        previous_token_name = Some(node_name);
    }

    // Add update events to delete the just added token
    let mut updates_delete = GraphUpdate::new();
    for t in token_names {
        updates_delete
            .add_event(UpdateEvent::DeleteNode {
                node_name: t.clone(),
            })
            .unwrap();
    }
    (updates_add, updates_delete)
}

fn apply_update_inmemory(bench: &mut Criterion) {
    if CORPUS_STORAGE.is_none() {
        return;
    }

    let cs = CORPUS_STORAGE.as_ref().unwrap();

    bench.bench_function("apply_update_inmemory", move |b| {
        cs.create_empty_corpus("apply_update_test_corpus", false)
            .unwrap();
        let (mut u1, mut u2) = create_example_update();
        b.iter(|| {
            cs.apply_update("apply_update_test_corpus", &mut u1)
                .unwrap();
            cs.apply_update("apply_update_test_corpus", &mut u2)
                .unwrap()
        });
        cs.delete("apply_update_test_corpus").unwrap();
    });
}

fn apply_update_ondisk(bench: &mut Criterion) {
    if CORPUS_STORAGE.is_none() {
        return;
    }

    let cs = CORPUS_STORAGE.as_ref().unwrap();

    bench.bench_function("apply_update_ondisk", move |b| {
        cs.create_empty_corpus("apply_update_test_corpus", true)
            .unwrap();
        let (mut u1, mut u2) = create_example_update();
        b.iter(|| {
            cs.apply_update("apply_update_test_corpus", &mut u1)
                .unwrap();
            cs.apply_update("apply_update_test_corpus", &mut u2)
                .unwrap()
        });
        cs.delete("apply_update_test_corpus").unwrap();
    });
}

criterion_group!(name=default; config= Criterion::default().sample_size(50); targets =
    deserialize_gum,
    find_first_ten_token_gum,
    find_first_ten_nouns_gum, 
    find_all_nouns_gum,
    count_extra_coref_anno, count_extra_node);
criterion_group!(name=apply_update; config= Criterion::default().sample_size(20); targets =
    apply_update_inmemory,
    apply_update_ondisk,
);
criterion_main!(default, apply_update);
