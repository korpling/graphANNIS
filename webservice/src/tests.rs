use std::{
    path::PathBuf,
    time::{Duration, SystemTime, UNIX_EPOCH},
};

use actix_web::{
    body::MessageBody,
    dev::{ServiceFactory, ServiceRequest, ServiceResponse},
    web, App,
};
use diesel::{r2d2::ConnectionManager, SqliteConnection};
use diesel_migrations::MigrationHarness;
use graphannis::{corpusstorage::ImportFormat, CorpusStorage};
use insta::assert_snapshot;
use jsonwebtoken::EncodingKey;
use log::{Level, Log, RecordBuilder};
use tempfile::NamedTempFile;

use crate::{
    api::administration::BackgroundJobs,
    auth::Claims,
    create_logger,
    settings::{JWTVerification, Settings},
};

pub const JWT_SECRET: &str = "not-a-secret";

pub fn create_empty_dbpool() -> r2d2::Pool<ConnectionManager<SqliteConnection>> {
    let manager = ConnectionManager::<SqliteConnection>::new(":memory:");
    let db_pool = r2d2::Pool::builder().build(manager).unwrap();
    let mut conn = db_pool.get().unwrap();
    conn.run_pending_migrations(crate::MIGRATIONS).unwrap();

    db_pool
}

pub fn create_test_app(
    cs: web::Data<CorpusStorage>,
    mut settings: Settings,
) -> App<
    impl ServiceFactory<
        ServiceRequest,
        Response = ServiceResponse<impl MessageBody>,
        Config = (),
        InitError = (),
        Error = actix_web::Error,
    >,
> {
    // Create an app that uses a string as secret so we can sign our own JWT
    // token.
    settings.auth.token_verification = JWTVerification::HS256 {
        secret: JWT_SECRET.to_string(),
    };

    let db_pool = create_empty_dbpool();

    let settings = web::Data::new(settings);
    let db_pool = web::Data::new(db_pool);
    let background_jobs = web::Data::new(BackgroundJobs::default());

    let app = crate::create_app(cs, settings, db_pool, background_jobs);
    app
}

pub fn create_auth_header() -> (&'static str, String) {
    // Create an auth header for an admin
    let in_sixty_minutes = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .checked_add(Duration::from_secs(3600))
        .unwrap();
    let admin_claims = Claims {
        sub: "admin".to_string(),
        exp: Some(in_sixty_minutes.as_millis() as i64),
        roles: vec!["admin".to_string()],
        groups: vec![],
    };
    let bearer_token = jsonwebtoken::encode(
        &jsonwebtoken::Header::default(),
        &admin_claims,
        &EncodingKey::from_secret(JWT_SECRET.as_ref()),
    )
    .unwrap();
    ("Authorization", format!("Bearer {bearer_token}"))
}

pub(crate) fn import_test_corpora(cs: &CorpusStorage) {
    let cargo_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    // Import three corpora A,B and C
    cs.import_from_fs(
        &cargo_dir.join("../graphannis/tests/SaltSampleCorpus.graphml"),
        ImportFormat::GraphML,
        Some("A".into()),
        false,
        true,
        |_| {},
    )
    .unwrap();

    cs.import_from_fs(
        &cargo_dir.join("../graphannis/tests/SaltSampleCorpus.graphml"),
        ImportFormat::GraphML,
        Some("B".into()),
        false,
        true,
        |_| {},
    )
    .unwrap();

    cs.import_from_fs(
        &cargo_dir.join("../graphannis/tests/SaltSampleCorpus.graphml"),
        ImportFormat::GraphML,
        Some("C".into()),
        false,
        true,
        |_| {},
    )
    .unwrap();
}

fn standard_filter() -> insta::Settings {
    let mut settings = insta::Settings::clone_current();
    // Remove any color ASCII codes
    settings.add_filter("\x1b", "");
    settings.add_filter("\\[[0-9]+m", "");

    // Filter out the time stamps
    settings.add_filter("[0-9]+:[0-9]+:[0-9]+ ", "12:00:00");
    // The loaded and also total available RAM size can vary
    settings.add_filter("[0-9.]+[MG]B / [0-9.]+[MG]B", "100MB / 300MB");
    // The loading and time can vary
    settings.add_filter("in [0-9]+ ms", "in 10 ms");
    settings
}

#[test]
fn test_logfile() -> Result<(), Box<dyn std::error::Error>> {
    let logfile = NamedTempFile::new()?;
    let mut settings = Settings::default();
    settings.logging.file = Some(logfile.path().to_string_lossy().to_string());

    // Get a logger
    let (logger, _) = create_logger(&settings)?;

    let record = RecordBuilder::new()
        .level(Level::Info)
        .args(format_args!("Hello World"))
        .build();
    logger.log(&record);

    let record = RecordBuilder::new()
        .level(Level::Debug)
        .args(format_args!("Debug Message"))
        .build();
    logger.log(&record);

    let logfile_content = std::fs::read_to_string(logfile.path())?;

    let snapshot_settings = standard_filter();
    snapshot_settings.bind(|| assert_snapshot!(logfile_content));

    Ok(())
}

#[test]
fn test_logfile_debug() -> Result<(), Box<dyn std::error::Error>> {
    let logfile = NamedTempFile::new()?;
    let mut settings = Settings::default();
    settings.logging.file = Some(logfile.path().to_string_lossy().to_string());
    settings.logging.debug = true;

    // Get a logger
    let (logger, _) = create_logger(&settings)?;

    let record = RecordBuilder::new()
        .level(Level::Info)
        .args(format_args!("Hello World"))
        .build();
    logger.log(&record);

    let record = RecordBuilder::new()
        .level(Level::Debug)
        .args(format_args!("Debug Message"))
        .build();
    logger.log(&record);

    let logfile_content = std::fs::read_to_string(logfile.path())?;

    let snapshot_settings = standard_filter();
    snapshot_settings.bind(|| assert_snapshot!(logfile_content));

    Ok(())
}
