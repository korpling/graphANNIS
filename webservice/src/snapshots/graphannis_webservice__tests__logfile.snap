---
source: webservice/src/tests.rs
expression: logfile_content
---
12:00:00[INFO] Hello World
